<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Codon Usage Analysis and Prediction of Gene Expressivity • coRdon</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/paper/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Codon Usage Analysis and Prediction of Gene Expressivity">
<meta property="og:description" content="Tool for analysis of codon usage in various 
    unannotated or KEGG/COG annotated DNA sequences. 
	Calculates different measures of CU bias and CU-based
	predictors of gene expressivity, and performs gene set 
	enrichment analysis for annotated sequences.
	Implements several methods for visualization of CU
	and enrichment analysis results.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">coRdon</a>
        <span class="label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased package">0.99.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/coRdon.html">How to</a>
</li>
<li>
  <a href="reference/index.html">Documentation</a>
</li>
<li>
  <a href="news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BioinfoHR/coRdon">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="http://bioinfo.hr">
    <span class="fa fa-globe fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    

    
    
<div id="cordon" class="section level2">
<h2 class="hasAnchor">
<a href="#cordon" class="anchor"></a>coRdon</h2>
<div id="codon-usage-analysis-prediction-of-gene-expressivity" class="section level4">
<h4 class="hasAnchor">
<a href="#codon-usage-analysis-prediction-of-gene-expressivity" class="anchor"></a>Codon Usage Analysis &amp; Prediction of Gene Expressivity</h4>
</div>
<div id="about" class="section level3">
<h3 class="hasAnchor">
<a href="#about" class="anchor"></a><i class="fa fa-list-ul" aria-hidden="true"></i> About</h3>
<p><br><strong>coRdon</strong> is an R package for analysis and visualization of codon usage in DNA sequences.</p>
<p>Main functionalities:</p>
<ul>
<li>calculates different measures of CU bias and CU-based predictors of gene expressivity<br>
</li>
<li>performs gene set enrichment analysis for unannotated or KEGG/COG annotated DNA sequences<br>
</li>
<li>implements several methods for visualization of codon usage and enrichment analysis results.</li>
</ul>
<p>Although specifically aimed at the analysis of metagenomic samples, coRdon allows for the inspection and quantification of codon usage in DNA sequences using any of the 20 different variants of genetic code, with the additional options to include stop codons and alternative start codons in calculations.</p>
<p>The following statistics are implemented in the package:</p>
<ul>
<li>ENC, effective number of codons (<a href="https://www.ncbi.nlm.nih.gov/pubmed/2110097">Wright, 1990</a>),<br>
</li>
<li>its modified version ENC’ (<a href="https://www.ncbi.nlm.nih.gov/pubmed/12140252">Novembre, 2002</a>),<br>
</li>
<li>a measure of codon bias, termed B (<a href="https://www.ncbi.nlm.nih.gov/pubmed/11489855">Karlin and Mrazek, 1996</a>),<br>
</li>
<li>and related measure of expression, E (<a href="https://www.ncbi.nlm.nih.gov/pubmed/10960111">Karlin and Mrazek, 2000</a>),<br>
</li>
<li>maximum likelihood codon bias, MCB (<a href="https://www.ncbi.nlm.nih.gov/pubmed/2110097">Urrutia and Hurst, 2001</a>),<br>
</li>
<li>MILC, Measure Independent of Length and Composition, and<br>
</li>
<li>MELP, MILC-based Expression Level Predictor (<a href="https://www.ncbi.nlm.nih.gov/pubmed/16029499">Supek and Vlahovicek, 2005</a>),</li>
<li>SCUO, synonymous codon usage orderliness (<a href="https://www.ncbi.nlm.nih.gov/pubmed/15222899">Wan et al., 2004</a>),</li>
<li>Codon Adaptation Index, CAI (<a href="https://www.ncbi.nlm.nih.gov/pubmed/3547335">Sharp and Li, 1987</a>),<br>
</li>
<li>frequency of optimal codons, Fop (<a href="https://www.ncbi.nlm.nih.gov/pubmed/6175758">Ikemura 1981</a>),<br>
</li>
<li>gene codon bias, GCB (<a href="http://www.ncbi.nlm.nih.gov/pubmed/14708578">Merkl, 2003</a>).</li>
</ul>
<p>The package also implements B plot for visualization of CU bias, both within a single sample and between different samples for which CU bias statistics are calculated.</p>
<p>Additionally, if the input sequences are annotated in either KEG or COG orthology database, functional analysis can be performed in order to determine significantly enriched functions in the imput sample. This is aimed particularly at metagenomic samples, as a way of determining functional fingerprint of a microbial community. There are also several methods for visualisation of enrichment analysis results, including MA-like plot and bar plot.</p>
<hr>
</div>
<div id="geting-started" class="section level3">
<h3 class="hasAnchor">
<a href="#geting-started" class="anchor"></a><i class="fa fa-cog" aria-hidden="true"></i> Geting started</h3>
<p><br> To install coRdon, run the following in R:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">source</span>(<span class="st">"https://bioconductor.org/biocLite.R"</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw">biocLite</span>(<span class="st">"coRdon"</span>)</a></code></pre></div>
<p>The developmental version can be installed directly from from GitHub:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"BioinfoHR/coRdon"</span>)</a></code></pre></div>
<p>For worked example on how to do analysis of codon usage with coRdon, please see <a href="articles/coRdon.html">the package vignette</a>.</p>
<hr>
</div>
<div id="citation" class="section level3">
<h3 class="hasAnchor">
<a href="#citation" class="anchor"></a><i class="fa fa-file" aria-hidden="true"></i> Citation</h3>
<p><br> TBA</p>
<p><i><img src="http://bioinfo.hr/wp-content/themes/theme1414/images/logo.png" alt="bioinfo.hr" title=""></i></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Report a bug at <br><a href="https://github.com/BioinfoHR/coRdon/issues">https://​github.com/​BioinfoHR/​coRdon/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/Artistic-2.0">Artistic-2.0</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="http://bioinfo.hr/members/">Anamaria Elek</a> <br><small class="roles"> Maintainer, author </small>  </li>
<li>
<a href="http://bioinfo.hr/maja-kuzman/">Maja Kuzman</a> <br><small class="roles"> Author </small>  </li>
<li>
<a href="http://bioinfo.hr/kristian-vlahovicek/">Kristian Vlahovicek</a> <br><small class="roles"> Author </small>  </li>
</ul>
</div>

      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="http://bioinfo.hr/members/">Anamaria Elek</a>, <a href="http://bioinfo.hr/maja-kuzman/">Maja Kuzman</a>, <a href="http://bioinfo.hr/kristian-vlahovicek/">Kristian Vlahovicek</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
