<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>coRdon • coRdon</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/paper/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="coRdon">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coRdon</a>
        <span class="label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased package">0.99.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/coRdon.html">How to</a>
</li>
<li>
  <a href="../reference/index.html">Documentation</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BioinfoHR/coRdon">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="http://bioinfo.hr">
    <span class="fa fa-globe fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>coRdon</h1>
                        <h4 class="author">Anamaria Elek</h4>
            
            <h4 class="date">2018-06-16</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BioinfoHR/coRdon/blob/master/vignettes/coRdon.Rmd"><code>vignettes/coRdon.Rmd</code></a></small>
      <div class="hidden name"><code>coRdon.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p><code>coRdon</code> provides tools for analysis of codone usage (CU) in various unannotated or KEGG/COG annotated DNA sequences. Funcionalities include: calculation of different CU bias statistics and CU-based predictors of gene expression, gene set enrichment analysis for annotated sequences, and several methods for visualization of CU and enrichment analysis results.</p>
    </div>
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Not all synonymous codons are used equally often in prokaryotic genomes - this selective preference is termed <strong>codon usage (CU) bias</strong>, and is an independent determinant of gene expression regulation at the translational level. Those synonymous codons corresponding to the most abundant tRNA species are considered optimal for translation. Codon usage bias can therfore be used to predict the relative expression levels of genes, by comparing CU bias of a gene to the CU bias of a set of genes known to be highly expressed.</p>
<p>This approach can be efficiently used to predict highly expressed genes in a single genome, but is especially useful at the higher level of an entire metagenome. It has been shown that, as well as being present within the genome, CU bias is shared among the microbial spieces in the same environment. By analysing CU bias of a metagenome, one can identify genes with high predicted expression across the entire microbial community, and determine which are the enriched functions within the community, i.e. its ‘functional fingerprint’.</p>
<p>A typical workflow for analysing CU bias includes:</p>
<ul>
<li>calculating one of the CU statistics</li>
<li>comparing CU of every sequence (ORF, gene) to average CU of a set of highly expressed genes</li>
<li>functional enrichment analysis for the subset of genes that have CU similar to a reference set of highly expressed genes, and are therefore predicted to be highly expressed themselves.</li>
</ul>
</div>
<div id="loading-sequences" class="section level1">
<h1 class="hasAnchor">
<a href="#loading-sequences" class="anchor"></a>Loading sequences</h1>
<p>Sequences from human gut microbiome samples of healthy individuals and liver cirrhosis patients (<a href="https://go.nature.com/2G7QdpC">Quin et al. 2014</a>) were processed, assembled and used to predict ORFs, which were then annotated with a KO (KEGG orthology) function (<a href="https://bit.ly/2DRfiz6">Fabijanic and Vlahovicek 2016</a>).</p>
<p>For each sequence, in each sample, we can calculate codon usage (CU) bias. In order to do this, we need to count occurrences of each codon in each sequence. This can be done by storing sequences in each sample as a <code>codonTable</code> object. We can use <code><a href="../reference/readSet.html">readSet()</a></code> to read (a directory containing) .fasta files and store sequences as a <code>DNAStringSet</code> object, which could then be converted to <code>codonTable</code> using the constructor method <code><a href="../reference/codonTable-class.html">codonTable()</a></code>.</p>
<p>For the purpose of this vignette we will create two <code>codonTable</code> objects, HD59 and LD94, containing codon counts for metagenomic gut sample from healthy individual, and from liver chirosis patient, respectively.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dnaLD94 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/readSet.html">readSet</a></span>(<span class="dt">file =</span> <span class="st">"http://hex.bioinfo.hr/~mfabijanic/LD94.fasta"</span>)
LD94 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/codonTable-class.html">codonTable</a></span>(dnaLD94)
dnaHD59 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/readSet.html">readSet</a></span>(<span class="dt">file =</span> <span class="st">"http://hex.bioinfo.hr/~mfabijanic/HD59.fasta"</span>)
HD59 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/codonTable-class.html">codonTable</a></span>(dnaHD59)</code></pre></div>
<p><code>codonTable</code> object stores codon counts for each sequence, as well as some additional metadata, namely sequence ID infered from a .fasta file or DNAStringSet object, sequence length in codons and KO / COG annotation. Note that not all of these need be present in every <code>codonTable</code> object - only the codon counts, and inferred lengths are mandatory. They can be accessed using <code><a href="../reference/codonTable-class.html">codonCounts()</a></code> and <code><a href="../reference/codonTable-class.html">getlen()</a></code> (not to be confused with <code>length()</code> which returns number of sequences in a <code>codonTable</code> object):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cc &lt;-<span class="st"> </span><span class="kw"><a href="../reference/codonTable-class.html">codonCounts</a></span>(HD59)
<span class="kw">head</span>(cc)</code></pre></div>
<pre><code>     AAA AAC AAG AAT ACA ACC ACG ACT AGA AGC AGG AGT ATA ATC ATG ATT CAA
[1,]   5   4  10   3   1   2   4   0   0   2   1   0   0   5   2   1   2
[2,]   3   5   4   1   2   3   5   1   0   1   0   0   0   8   4   4   0
[3,]   4   3   5   5   1   5   8   0   0   1   1   0   0  13   8   4   2
[4,]   1   9  12   3   1  14   7   0   0   5   0   0   0  15   6   6   2
[5,]   5   2   3   2   0   0   1   0   0   0   0   0   0   3   1   4   0
[6,]   5   1   2   3   3   5   1   3   1   1   0   4   5   7  10   4   3
     CAC CAG CAT CCA CCC CCG CCT CGA CGC CGG CGT CTA CTC CTG CTT GAA GAC
[1,]   1   9   2   0   2   7   0   1   3   7   0   0  12  15   2  11   4
[2,]   4   3   1   0   1   4   1   0   4   1   0   0   5  11   2   3   9
[3,]   4  16   7   1   2  12   2   1   7   0   9   0   4  14   4   7  13
[4,]   2  15   3   0   1   4   1   0   9   1   8   0   5  15   1  10  12
[5,]   1   1   1   0   1   0   0   0   1   1   0   0   1   4   0   1   3
[6,]   4   6   0   0   7   1   3   0   5   0   3   0   2   6   0  16   4
     GAG GAT GCA GCC GCG GCT GGA GGC GGG GGT GTA GTC GTG GTT TAA TAC TAG
[1,]   5   5   2   8   6   0   1   2   2   2   2   4   2   2   0   1   0
[2,]   9   2   2   7  13   1   2  12   4   2   0   9   3   2   0   1   0
[3,]   8   5   4  10   9   2   1  17   1   3   1   8  12   2   0   6   0
[4,]   8   7   3   5   8   0   2   9   0   1   0   3  12   2   0   6   0
[5,]   4   1   1   8   2   1   0   4   0   1   0   0   2   0   0   0   0
[6,]   1   3   8  11   1   4   7   8   0   8   7   6   3   2   0   3   0
     TAT TCA TCC TCG TCT TGA TGC TGG TGT TTA TTC TTG TTT
[1,]   5   0   4   1   3   0   6   1   1   0   3   1   5
[2,]   5   1   4   6   0   0   4   0   0   1   2   2   2
[3,]   3   0  10   5   0   0   1   1   0   0   7   8   0
[4,]   1   0   5   3   0   0   0   0   0   0  10   3   3
[5,]   0   0   2   4   0   0   0   1   0   0   0   1   1
[6,]   4   2   2   2   0   0   1   2   1   3   3   3   2</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">l &lt;-<span class="st"> </span><span class="kw"><a href="../reference/codonTable-class.html">getlen</a></span>(HD59)
<span class="kw">head</span>(l)</code></pre></div>
<pre><code>[1] 192 186 287 259  69 212</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(HD59)</code></pre></div>
<pre><code>[1] 19161</code></pre>
<p>We can get IDs, KO or COG annotations in a similar way, if these are present:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ko &lt;-<span class="st"> </span><span class="kw"><a href="../reference/codonTable-class.html">getKO</a></span>(HD59)
<span class="kw">head</span>(ko)</code></pre></div>
<pre><code>[1] "K01509" "K03431" "K08884" "K02313" "K01810" "K00174"</code></pre>
</div>
<div id="calculate-cu-bias" class="section level1">
<h1 class="hasAnchor">
<a href="#calculate-cu-bias" class="anchor"></a>Calculate CU bias</h1>
<p>Now we can calculate CU bias for every sequence. There are many statistics that measure codon usage, and several widely used ones are implemented in coRdon. For example, Measure Independent of Length and Composition (MILC) can be calculated for sequences in the <code>codonTable</code> like so:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">milc &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MILC.html">MILC</a></span>(HD59)
<span class="kw">head</span>(milc)</code></pre></div>
<pre><code>          self
[1,] 0.7231937
[2,] 0.7083612
[3,] 0.7014576
[4,] 0.7119265
[5,] 0.5658703
[6,] 0.7242135</code></pre>
<p>By default, MILC value for every sequence in a set is calculated with respect to the average CU bias of the entire sample (“self”). This can be changed by setting <code>self = FALSE</code> and providing a subset of reference genes, to which CU will be compared. This is commonly a subset of genes known to be highly expressed, e.g. ribosomal genes. If the sequences in the set are annotated, we can choose to calculate CU bias with respect to codon usage of ribosomal genes (i.e. codon usage of the subset of sequences in the sample which are annotated as ribosomal genes) simply by setting <code>ribosomal = TRUE</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">milc &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MILC.html">MILC</a></span>(HD59, <span class="dt">ribosomal =</span> <span class="ot">TRUE</span>)
<span class="kw">head</span>(milc)</code></pre></div>
<pre><code>          self ribosomal
[1,] 0.7231937 1.0337208
[2,] 0.7083612 0.9530175
[3,] 0.7014576 0.8365788
[4,] 0.7119265 0.7975796
[5,] 0.5658703 0.8343067
[6,] 0.7242135 0.7887504</code></pre>
<p>Alternatively, instead of using ribosomal genes, we could use a <code>subsets</code> argument to provide a named list of character vectors of annotations for different genes to be used as reference set(s). However, if no annotation for sequences is available, one can still calculate CU bias with respect to some chosen subset(s) of genes, by setting <code>subsets</code> to be a named list of logical vectors of sequences to be included in this reference subset(s), or a named list with an entirely new reference codon counts table(s).</p>
<p>Also of note, <code>self</code>, <code>ribosomal</code> and <code>subsets</code> arguments can be combined, and MILC calculated for all sequences, with respect to each of thus specified reference sets of genes.</p>
<p>Another important consideration is that MILC, like most of the other CU bias statistics (and contrary to its name), is still somewhat dependent on the length of sequence for which it is calculated, and in order to obtain meaningful values, it is recommended that sequences should be at least 80 codons long. To that end, either soft or hard filtering step can be included in calculation of CU statistics. Hard filtering will remove all of the sequences from <code>codonTable</code> that are shorter than some specified threshold (length in codons, default is 80), whereas soft filtering will produce a message if there are such sequences in the set, but will not remove them.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">milc &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MILC.html">MILC</a></span>(HD59, <span class="dt">filtering =</span> <span class="st">"soft"</span>)</code></pre></div>
<pre><code>Warning in MILC(HD59, filtering = "soft"): Some sequences have below-
threshold length!</code></pre>
<p>We can visually inspect distribution of sequence lengths, extracted from <code>codonTable</code> object using <code><a href="../reference/codonTable-class.html">getlen()</a></code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">lengths &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw"><a href="../reference/codonTable-class.html">getlen</a></span>(HD59))
<span class="kw">colnames</span>(lengths) &lt;-<span class="st"> "length"</span>
<span class="kw">ggplot</span>(lengths, <span class="kw">aes</span>(length)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_density</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> <span class="dv">80</span>, <span class="dt">colour =</span> <span class="st">"red"</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme_light</span>()</code></pre></div>
<p><img src="coRdon_files/figure-html/unnamed-chunk-8-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Now we calculate MILC values with respect to overall sample CU and ribosomal genes CU, removing beforehand sequences shorter than 80 codons (use <code>len.threshold</code> argument to specify differnt value of threshold):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">milc &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MILC.html">MILC</a></span>(HD59, <span class="dt">ribosomal =</span> <span class="ot">TRUE</span>, <span class="dt">filtering =</span> <span class="st">"hard"</span>)</code></pre></div>
<p>Other CU statistics can be calculated in the same way as <code><a href="../reference/MILC.html">MILC()</a></code>, using one of the functions: <code><a href="../reference/B.html">B()</a></code>, <code><a href="../reference/MCB.html">MCB()</a></code>, <code><a href="../reference/ENCprime.html">ENCprime()</a></code>, <code><a href="../reference/ENC.html">ENC()</a></code> or <code><a href="../reference/SCUO.html">SCUO()</a></code>. Note however, that when calculating ENC and SCUO, one doesn’t need to provide a subset of reference genes, because these statistics measure distance in codon usage to uniform usage of synonymous codons.</p>
<p>Some additional arguments that can be set when calculating CU statistics are <code>id_or_name2</code>, for choosing a genetic code variant (see help for <code>getGeneticCode()</code> function in Biostrings package), <code>alt.init</code> for alternative initiation codons, <code>stop.rm</code> for removal of STOP codons.</p>
<div id="visualisation-of-cu-bias" class="section level2">
<h2 class="hasAnchor">
<a href="#visualisation-of-cu-bias" class="anchor"></a>Visualisation of CU bias</h2>
<p>Now we can visualize CU bias for every gene on the B plot.</p>
<p>Here, every gene is represented by a single point on the plot, its coordinates determined by distance of genes’ CU bias to overall CU bias (“self”, on y axis) and to CU bias of reference genes (“ribosomal”, on x axis).<br><code>x</code> and <code>y</code> are names of columns in a matrix given as <code>data</code> argument, they can alternatevly be numeric vectors (of the same length), containing values of the CU statistic we wish to plot one against the other.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)

xlab &lt;-<span class="st"> "MILC distance from sample centroid"</span>
ylab &lt;-<span class="st"> "MILC distance from ribosomal genes"</span>

milc_HD59 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MILC.html">MILC</a></span>(HD59, <span class="dt">ribosomal =</span> <span class="ot">TRUE</span>)
<span class="kw"><a href="../reference/Bplot.html">Bplot</a></span>(<span class="dt">x =</span> <span class="st">"ribosomal"</span>, <span class="dt">y =</span> <span class="st">"self"</span>, <span class="dt">data =</span> milc_HD59) <span class="op">+</span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> xlab, <span class="dt">y =</span> ylab)</code></pre></div>
<p><img src="coRdon_files/figure-html/unnamed-chunk-10-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">milc_LD94 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MILC.html">MILC</a></span>(LD94, <span class="dt">ribosomal =</span> <span class="ot">TRUE</span>)
<span class="kw"><a href="../reference/Bplot.html">Bplot</a></span>(<span class="dt">x =</span> <span class="st">"ribosomal"</span>, <span class="dt">y =</span> <span class="st">"self"</span>, <span class="dt">data =</span> milc_LD94) <span class="op">+</span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> xlab, <span class="dt">y =</span> ylab)</code></pre></div>
<p>The argument <code>annotations</code> needs to be specified when we wish to indicate certain genes on the plot. For example, to indicate ribosomal genes, provide character vector of annotations corresponding to values in <code>x</code>, <code>y</code> and <code>data</code>, and set <code>ribosomal = TRUE</code>. Note that, if we filtered sequences when calculating MILC, we need to provide the annotations for those sequences only.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/codonTable-class.html">getKO</a></span>(HD59)[<span class="kw"><a href="../reference/codonTable-class.html">getlen</a></span>(HD59) <span class="op">&gt;</span><span class="st"> </span><span class="dv">80</span>]
<span class="kw"><a href="../reference/Bplot.html">Bplot</a></span>(<span class="dt">x =</span> <span class="st">"ribosomal"</span>, <span class="dt">y =</span> <span class="st">"self"</span>, <span class="dt">data =</span> milc,
      <span class="dt">annotations =</span> genes, <span class="dt">ribosomal =</span> <span class="ot">TRUE</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> xlab, <span class="dt">y =</span> ylab)</code></pre></div>
<p><img src="coRdon_files/figure-html/unnamed-chunk-12-1.png" width="700" style="display: block; margin: auto;"></p>
<p>To indicate any other gene(s), use <code>reference</code> argument. <code>reference</code> takes a list of length 1, containing either a logical vector in which <code>TRUE</code> corresponds (by position) to genes that are to be indicated on the plot, and <code>FALSE</code> to all the other genes, or a character vector of annotations for the genes to be indicated on the plot.</p>
<p>Another interesting way to visualise codon usage is to plot the CU distances between the two samples on the B plot. This can be done using <code><a href="../reference/intraBplot.html">intraBplot()</a></code> function which takes as arguments two <code>codonTable</code> objects and a character indicating which CU statistic to plot. We can also choose to indicate ribosomal genes on the plot by setting <code>ribosomal = TRUE</code>, these genes will now be shown as stronger points on the plot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/intraBplot.html">intraBplot</a></span>(HD59, LD94, <span class="dt">names =</span> <span class="kw">c</span>(<span class="st">"HD59"</span>, <span class="st">"LD94"</span>), 
           <span class="dt">variable =</span> <span class="st">"MILC"</span>, 
           <span class="dt">ribosomal =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p><img src="coRdon_files/figure-html/unnamed-chunk-13-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
</div>
<div id="predict-genes-expressivity" class="section level1">
<h1 class="hasAnchor">
<a href="#predict-genes-expressivity" class="anchor"></a>Predict genes’ expressivity</h1>
<p>Next, we predict expression levels of genes in each sample, with ribosomal genes once again used as a reference. There are several measures of CU-based gene expressivity implemented in coRdon. Here we calculate values of MILC-based Expression Level Predictor (MELP), with respect to ribosomal genes, and excluding sequencs shorter than 80 codons (this is the default value of <code>len.threshold</code>):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">melp &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MELP.html">MELP</a></span>(HD59, <span class="dt">ribosomal =</span> <span class="ot">TRUE</span>, <span class="dt">filtering =</span> <span class="st">"hard"</span>)
<span class="kw">head</span>(melp)</code></pre></div>
<pre><code>     ribosomal
[1,] 0.7112424
[2,] 0.7514472
[3,] 0.8478242
[4,] 0.9008746
[5,] 0.9106633
[6,] 0.8318351</code></pre>
<p>Other statistics that measure gene expresivity can be calculated analogously, using <code><a href="../reference/E.html">E()</a></code>, <code><a href="../reference/CAI.html">CAI()</a></code>, <code><a href="../reference/GCB.html">GCB()</a></code>, and <code><a href="../reference/Fop.html">Fop()</a></code> functions.</p>
<p>Genes from a single sample with high expressivity values (e.g. MILC greater than 1) are considered to be optimized for translation in that sample. If annotation for genes is available, we can perform enrichment analysis in order to determine which functions are significantly enriched in any given sample.</p>
</div>
<div id="functional-annotation" class="section level1">
<h1 class="hasAnchor">
<a href="#functional-annotation" class="anchor"></a>Functional annotation</h1>
<p>We can identify most significantly enriched or depleted functions in the set of annotated genes predicted to have high expression level.</p>
<p>For a single sample, we first create a contingency table summarising counts of genes annotated to each KO category among all the genes in sample, and among those predicted to be highly expressed. We do this using <code><a href="../reference/crossTab-class.html">crossTab()</a></code> function, giving it a character vector of genes’ annotations, and a numeric vector of their respective MELP values, and specifying that a subset of highly expressed reference genes should contain those genes that have MELP value grater than 1 (this is default value of threshold).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ct &lt;-<span class="st"> </span><span class="kw"><a href="../reference/crossTab-class.html">crossTab</a></span>(genes, <span class="kw">as.numeric</span>(melp), <span class="dt">threshold =</span> 1L)
ct</code></pre></div>
<pre><code>crossTab instance for 17829 sequences.
Sequence annotations:
  chr [1:17829] "K01509" "K03431" "K08884" "K02313" "K00174" "K01846" ... 
Corresponding variable values:
  num [1:17829] 0.711 0.751 0.848 0.901 0.911 ... 
Contingecy table:
      category all gt_1
   1:   K00001  16    4
   2:   K00002   1    0
   3:   K00003  16    4
   4:   K00005   5    1
   5:   K00008   1    1
  ---                  
1660:   K13789  17    1
1661:   K13821   3    0
1662:   K13829   2    0
1663:   K13923   1    0
1664:   K13924   3    0</code></pre>
<p>Output of <code><a href="../reference/crossTab-class.html">crossTab()</a></code> function is an object of <code>crossTab</code> class, containing gene anotations, respective values of the given variable, and a contingency table with counts for all genes, and for each defined subset of genes. They can all be accessed like so:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/crossTab-class.html">contable</a></span>(ct)</code></pre></div>
<pre><code>      category all gt_1
   1:   K00001  16    4
   2:   K00002   1    0
   3:   K00003  16    4
   4:   K00005   5    1
   5:   K00008   1    1
  ---                  
1660:   K13789  17    1
1661:   K13821   3    0
1662:   K13829   2    0
1663:   K13923   1    0
1664:   K13924   3    0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ann &lt;-<span class="st"> </span><span class="kw"><a href="../reference/crossTab-class.html">getSeqAnnot</a></span>(ct)
<span class="kw">head</span>(ann)</code></pre></div>
<pre><code>[1] "K01509" "K03431" "K08884" "K02313" "K00174" "K01846"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">var &lt;-<span class="st"> </span><span class="kw"><a href="../reference/crossTab-class.html">getVariable</a></span>(ct)
<span class="kw">head</span>(var)</code></pre></div>
<pre><code>[1] 0.7112424 0.7514472 0.8478242 0.9008746 0.9106633 0.8318351</code></pre>
<p>We could have also defined the subset of highly expressed genes by specifying a percent of genes with highest MELP values:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/crossTab-class.html">crossTab</a></span>(genes, <span class="kw">as.numeric</span>(melp), <span class="dt">percentiles =</span> <span class="fl">0.05</span>)</code></pre></div>
<pre><code>crossTab instance for 17829 sequences.
Sequence annotations:
  chr [1:17829] "K01509" "K03431" "K08884" "K02313" "K00174" "K01846" ... 
Corresponding variable values:
  num [1:17829] 0.711 0.751 0.848 0.901 0.911 ... 
Contingecy table:
      category all top_0.05 gt_1
   1:   K00001  16        0    4
   2:   K00002   1        0    0
   3:   K00003  16        2    4
   4:   K00005   5        0    1
   5:   K00008   1        1    1
  ---                           
1660:   K13789  17        0    1
1661:   K13821   3        0    0
1662:   K13829   2        0    0
1663:   K13923   1        0    0
1664:   K13924   3        0    0</code></pre>
<p>Of note, both <code>threshold</code> and <code>percentiles</code> can be numerical vectors, in which case the functions used in subsequent analysis (<code>enrichment</code> and associated plotting functions, <code>enrichMAplot</code>, <code>enrichBarplot</code>) will produce output for each thus specified reference susbset.</p>
<p>Having a contingency table, we can perform enrichment analysis. This implies scaling and transforming gene counts by MA transformation, and performing binomial test, optinally inncluding correction for multiple testing, as chosen by specifying the <code>pAdjustMethod</code> argument (see <code>p.adjust.methods</code> for possible options). Additional parameters <code>pvalueCutoff</code> and <code>padjCutoff</code> can be passed to the <code><a href="../reference/enrichment.html">enrichment()</a></code> function in order to exclude from the results those sequences that have higher significance levels than specified.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">enr &lt;-<span class="st"> </span><span class="kw"><a href="../reference/enrichment.html">enrichment</a></span>(ct)
enr</code></pre></div>
<pre><code>An object of class 'AnnotatedDataFrame'
  rowNames: 1 2 ... 1664 (1664 total)
  varLabels: category all ... padj (8 total)
  varMetadata: labelDescription</code></pre>
<p>The output is an object of <code>AnnotatedDataFrame</code> class from Biobase package. The data stored in this object can be accesed by using <code>pData()</code> method defined in Biobase.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(Biobase)
enr_data &lt;-<span class="st"> </span><span class="kw">pData</span>(enr)
<span class="kw">head</span>(enr_data)</code></pre></div>
<pre><code>  category all gt_1     enrich          M         A     pvals      padj
1   K00001  16    4  34.438764  0.4269492 2.1084535 0.3559590 0.9631151
2   K00002   1    0 -14.526104 -0.2264442 0.1132221 1.0000000 1.0000000
3   K00003  16    4  34.438764  0.4269492 2.1084535 0.3559590 0.9631151
4   K00005   5    1   8.123351  0.1126781 0.9436609 0.5725245 1.0000000
5   K00008   1    1  70.947792  0.7735558 0.6132221 0.1562952 0.6095665
6   K00009   2    0 -25.367324 -0.4221207 0.2110603 1.0000000 1.0000000</code></pre>
<div id="visualisation-of-enrichment" class="section level2">
<h2 class="hasAnchor">
<a href="#visualisation-of-enrichment" class="anchor"></a>Visualisation of enrichment</h2>
<p>We can plot enriched and depleted KO categories on an MA plot, with significant enrichment or depletion defined as those having the p-value below the level of significance <code>siglev</code> 0.05 (one can also use adjusted p-value, by setting <code>pvalue</code> = “padj”, and different values of <code>siglev</code>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/enrichMAplot.html">enrichMAplot</a></span>(enr, <span class="dt">pvalue =</span> <span class="st">"pvals"</span>, <span class="dt">siglev =</span> <span class="fl">0.05</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme_light</span>()</code></pre></div>
<p><img src="coRdon_files/figure-html/unnamed-chunk-20-1.png" width="700" style="display: block; margin: auto;"></p>
<p>In order to determine functions that are enriched among highly expressed genes, we need to map KO annotations to a broader onthology. This can be done using <code><a href="../reference/reduceCrossTab.html">reduceCrossTab()</a></code> method which takes a <code>crossTab</code> object as its first argument, and maps KO categories to either KEGG Pathway, KEGG Module, or COG functional categories, depending on the value of <code>target</code> argument.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ctpath &lt;-<span class="st"> </span><span class="kw"><a href="../reference/reduceCrossTab.html">reduceCrossTab</a></span>(ct, <span class="dt">target =</span> <span class="st">"pathway"</span>)
ctpath</code></pre></div>
<pre><code>crossTab instance for 17829 sequences.
Sequence annotations:
  chr [1:17829] "K01509" "K03431" "K08884" "K02313" "K00174" "K01846" ... 
Corresponding variable values:
  num [1:17829] 0.711 0.751 0.848 0.901 0.911 ... 
Contingecy table:
     category all gt_1
  1: map00010 515  187
  2: map00020 257  107
  3: map00030 321  104
  4: map00040 212   61
  5: map00051 301   74
 ---                  
255: map05215  23    5
256: map05230 141   38
257: map05231   6    0
258: map05340  15    0
259: map05418  68   13</code></pre>
<p>We can then perform enrichment analysis using the reduced contingency table.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">enrpath &lt;-<span class="st"> </span><span class="kw"><a href="../reference/enrichment.html">enrichment</a></span>(ctpath)
enrpath_data &lt;-<span class="st"> </span><span class="kw">pData</span>(enrpath)
<span class="kw">head</span>(enrpath_data)</code></pre></div>
<pre><code>  category all gt_1    enrich          M        A        pvals
1 map00010 515  187  66.05321  0.7316456 7.188766 8.254035e-11
2 map00020 257  107  89.47564  0.9220124 6.293881 1.073265e-09
3 map00030 321  104  48.00241  0.5656207 6.431435 1.174059e-04
4 map00040 212   61  31.37244  0.3936627 5.757365 3.233125e-02
5 map00051 301   74  12.63483  0.1716531 6.142992 2.918974e-01
6 map00052 476   52 -49.38825 -0.9824556 6.219148 2.774278e-08
          padj
1 1.068897e-09
2 1.208590e-08
3 5.737381e-04
4 9.968802e-02
5 5.446909e-01
6 2.395127e-07</code></pre>
<p>Functional enrichment results can also be plotted on a bar plot, using enrichBarplot() method. Here, bars are coloured by the level of significance, <code>padj</code>, whereas their height is the relative enrichment <code>"enrich"</code> (can also be mean average of scaled counts, <code>"A"</code>, or scaled counts ratio, <code>"M"</code>). If we also specify <code>siglev</code>, only those categories for which associated <code>pvalue</code> (“pvals” or “padj”) is below the given value will be ploted.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/enrichBarplot.html">enrichBarplot</a></span>(enrpath, <span class="dt">variable =</span> <span class="st">"enrich"</span>, 
              <span class="dt">pvalue =</span> <span class="st">"padj"</span>, <span class="dt">siglev =</span> <span class="fl">0.05</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme_light</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">coord_flip</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">"category"</span>, <span class="dt">y =</span> <span class="st">"relative enrichment"</span>)</code></pre></div>
<p><img src="coRdon_files/figure-html/unnamed-chunk-23-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Prior to plotting, KEGGREST package can be used to match pathway identifiers to pathway descriptions, in order to produce a more informative plot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(KEGGREST)
paths &lt;-<span class="st"> </span><span class="kw">names</span>(<span class="kw">keggList</span>(<span class="st">"pathway"</span>))
paths &lt;-<span class="st"> </span><span class="kw">regmatches</span>(paths, <span class="kw">regexpr</span>(<span class="st">"[[:alpha:]]{2,4}</span><span class="ch">\\</span><span class="st">d{5}"</span>, paths))
pnames &lt;-<span class="st"> </span><span class="kw">unname</span>(<span class="kw">keggList</span>(<span class="st">"pathway"</span>))
ids &lt;-<span class="st"> </span><span class="kw">match</span>(<span class="kw">pData</span>(enrpath)<span class="op">$</span>category, paths)
descriptions &lt;-<span class="st"> </span>pnames[ids]
<span class="kw">pData</span>(enrpath)<span class="op">$</span>category &lt;-<span class="st"> </span>descriptions
enrpath_data &lt;-<span class="st"> </span><span class="kw">pData</span>(enrpath)</code></pre></div>
</div>
</div>
<div id="integration" class="section level1">
<h1 class="hasAnchor">
<a href="#integration" class="anchor"></a>Integration</h1>
<p>In order to pipe the results from CU analysis in various downstream applications, it is useful to have the data from different samples in a single matrix, with sequences (genes) in rows and values for different samples in columns. This can be achieved with <code><a href="../reference/enrich.matrix.html">enrich.matrix()</a></code> function.</p>
<p>For example, if we performed CU-based enrichment analysis on the two metagenomic samples described at the beginning of this vignette, and wanted to see how the results compare by plotting them on a heatmap (e.g. using the ComplexHeatmap package), we could do the following:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># calculate MELP</span>
melpHD59 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MELP.html">MELP</a></span>(HD59, <span class="dt">ribosomal =</span> <span class="ot">TRUE</span>, 
                 <span class="dt">filtering =</span> <span class="st">"hard"</span>, <span class="dt">len.threshold =</span> <span class="dv">100</span>)
genesHD59 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/codonTable-class.html">getKO</a></span>(HD59)[<span class="kw"><a href="../reference/codonTable-class.html">getlen</a></span>(HD59) <span class="op">&gt;</span><span class="st"> </span><span class="dv">100</span>]

melpLD94 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MELP.html">MELP</a></span>(LD94, <span class="dt">ribosomal =</span> <span class="ot">TRUE</span>, 
                 <span class="dt">filtering =</span> <span class="st">"hard"</span>, <span class="dt">len.threshold =</span> <span class="dv">100</span>)
genesLD94 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/codonTable-class.html">getKO</a></span>(LD94)[<span class="kw"><a href="../reference/codonTable-class.html">getlen</a></span>(LD94) <span class="op">&gt;</span><span class="st"> </span><span class="dv">100</span>]

<span class="co"># make cntingency table</span>
ctHD59 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/crossTab-class.html">crossTab</a></span>(genesHD59, <span class="kw">as.numeric</span>(melpHD59))
ctLD94 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/crossTab-class.html">crossTab</a></span>(genesLD94, <span class="kw">as.numeric</span>(melpLD94))

ctHD59 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/reduceCrossTab.html">reduceCrossTab</a></span>(ctHD59, <span class="st">"pathway"</span>)
ctLD94 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/reduceCrossTab.html">reduceCrossTab</a></span>(ctLD94, <span class="st">"pathway"</span>)

<span class="co"># calculate enrichment</span>
enrHD59 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/enrichment.html">enrichment</a></span>(ctHD59)
enrLD94 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/enrichment.html">enrichment</a></span>(ctLD94)

mat &lt;-<span class="st"> </span><span class="kw"><a href="../reference/enrich.matrix.html">enrich.matrix</a></span>(<span class="kw">list</span>(<span class="dt">HD59 =</span> enrHD59, <span class="dt">LD94 =</span> enrLD94), 
                     <span class="dt">variable =</span> <span class="st">"enrich"</span>)
<span class="kw">head</span>(mat)</code></pre></div>
<pre><code>              HD59      LD94
map00010  65.39609  89.29378
map00020  92.37626  92.37498
map00030  52.80571  64.70925
map00040  22.14637  11.64027
map00051  16.61736  23.71068
map00052 -52.22516 -58.08676</code></pre>
<p>Thus created matrix, with genes in rows and samples in columns, can be directly used as an argument to <code>Heatmap()</code> function for plotting. Also, for the purpose of neat visualization, we can again make use of the KEGGREST package to match pathway identifiers to pathway descriptions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">paths &lt;-<span class="st"> </span><span class="kw">names</span>(KEGGREST<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/KEGGREST/topics/keggList">keggList</a></span>(<span class="st">"pathway"</span>))
paths &lt;-<span class="st"> </span><span class="kw">regmatches</span>(paths, <span class="kw">regexpr</span>(<span class="st">"[[:alpha:]]{2,4}</span><span class="ch">\\</span><span class="st">d{5}"</span>, paths))
pnames &lt;-<span class="st"> </span><span class="kw">unname</span>(KEGGREST<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/KEGGREST/topics/keggList">keggList</a></span>(<span class="st">"pathway"</span>))
ids &lt;-<span class="st"> </span><span class="kw">match</span>(<span class="kw">rownames</span>(mat), paths)
descriptions &lt;-<span class="st"> </span>pnames[ids]
<span class="kw">rownames</span>(mat) &lt;-<span class="st"> </span>descriptions

mat &lt;-<span class="st"> </span>mat[<span class="kw">apply</span>(mat, <span class="dv">1</span>, <span class="cf">function</span>(x) <span class="kw">all</span>(x<span class="op">!=</span><span class="dv">0</span>)), ]
ComplexHeatmap<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/ComplexHeatmap/topics/Heatmap">Heatmap</a></span>(
    mat, 
    <span class="dt">name =</span> <span class="st">"relative </span><span class="ch">\n</span><span class="st">enrichment"</span>,
    <span class="dt">col =</span> circlize<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/circlize/topics/colorRamp2">colorRamp2</a></span>(<span class="kw">c</span>(<span class="op">-</span><span class="dv">100</span>, <span class="dv">0</span>, <span class="dv">100</span>), 
                               <span class="kw">c</span>(<span class="st">"red"</span>, <span class="st">"white"</span>, <span class="st">"blue"</span>)),
    <span class="dt">row_names_side =</span> <span class="st">"left"</span>,
    <span class="dt">row_names_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>),
    <span class="dt">show_column_dend =</span> <span class="ot">FALSE</span>, 
    <span class="dt">show_row_dend =</span> <span class="ot">FALSE</span>)</code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#loading-sequences">Loading sequences</a></li>
      <li>
<a href="#calculate-cu-bias">Calculate CU bias</a><ul class="nav nav-pills nav-stacked">
<li><a href="#visualisation-of-cu-bias">Visualisation of CU bias</a></li>
      </ul>
</li>
      <li><a href="#predict-genes-expressivity">Predict genes’ expressivity</a></li>
      <li>
<a href="#functional-annotation">Functional annotation</a><ul class="nav nav-pills nav-stacked">
<li><a href="#visualisation-of-enrichment">Visualisation of enrichment</a></li>
      </ul>
</li>
      <li><a href="#integration">Integration</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="http://bioinfo.hr/members/">Anamaria Elek</a>, <a href="http://bioinfo.hr/maja-kuzman/">Maja Kuzman</a>, <a href="http://bioinfo.hr/kristian-vlahovicek/">Kristian Vlahovicek</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
