<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Calculate gene expression measure (E). — E • coRdon</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/paper/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Calculate gene expression measure (E). — E" />

<meta property="og:description" content="Calculate E values for every sequence in the given
codonTable object." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coRdon</a>
        <span class="label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased package">0.99.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/coRdon.html">How to</a>
</li>
<li>
  <a href="../reference/index.html">Documentation</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/BioinfoHR/coRdon">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="http://bioinfo.hr">
    <span class="fa fa-globe fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate gene expression measure (E).</h1>
    <small class="dont-index">Source: <a href='https://github.com/BioinfoHR/coRdon/blob/master/R/codonUsage-expressivity.R'><code>R/codonUsage-expressivity.R</code></a></small>
    <div class="hidden name"><code>E.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Calculate E values for every sequence in the given
<code>codonTable</code> object.</p>
    
    </div>

    <pre class="usage"><span class='fu'>E</span>(<span class='no'>cTobject</span>, <span class='kw'>subsets</span> <span class='kw'>=</span> <span class='fu'>list</span>(), <span class='kw'>ribosomal</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>id_or_name2</span> <span class='kw'>=</span> <span class='st'>"1"</span>,
  <span class='kw'>alt.init</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>stop.rm</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>filtering</span> <span class='kw'>=</span> <span class='st'>"none"</span>,
  <span class='kw'>len.threshold</span> <span class='kw'>=</span> <span class='fl'>80</span>)

<span class='co'># S4 method for codonTable</span>
<span class='fu'>E</span>(<span class='no'>cTobject</span>, <span class='kw'>subsets</span> <span class='kw'>=</span> <span class='fu'>list</span>(), <span class='kw'>ribosomal</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>id_or_name2</span> <span class='kw'>=</span> <span class='st'>"1"</span>, <span class='kw'>alt.init</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>stop.rm</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>filtering</span> <span class='kw'>=</span> <span class='st'>"none"</span>, <span class='kw'>len.threshold</span> <span class='kw'>=</span> <span class='fl'>80</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>cTobject</th>
      <td><p>A <code>codonTable</code> object.</p></td>
    </tr>
    <tr>
      <th>subsets</th>
      <td><p>A (named) list of logical vectors, the length of each equal
to <code><a href='codonTable-class.html'>getlen(cTobject)</a></code>, i.e. the number of sequences in the set, or
character vectors (of any length) containing KEGG/eggNOG annotations,
or codonTable objects (of any length).</p></td>
    </tr>
    <tr>
      <th>ribosomal</th>
      <td><p>Logical, if <code>TRUE</code>, CU statistic is also calculated
against the average CU of the ribosomal genes in the sequence set.</p></td>
    </tr>
    <tr>
      <th>id_or_name2</th>
      <td><p>A single string that uniquely identifies the genetic code to extract.
    Should be one of the values in the <code>id</code> or <code>name2</code> columns
    of <code>GENETIC_CODE_TABLE</code>.</p></td>
    </tr>
    <tr>
      <th>alt.init</th>
      <td><p>logical, whether to use alternative initiation codons.
Default is <code>TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>stop.rm</th>
      <td><p>Logical, whether to remove stop codons. Default is
<code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>filtering</th>
      <td><p>Character vector, one of <code>c("none", "soft", "hard")</code>.
Specifies whether sequences shorther than some threshold value of
length (in codons), <code>len.threshold</code>, should be excluded from
calculations. If <code>"none"</code> (default), length of sequences is not
checked, if <code>"soft"</code>, a warrning is printed if there are shorter
sequences, and if <code>"hard"</code>, these sequences are excluded from
calculation.</p></td>
    </tr>
    <tr>
      <th>len.threshold</th>
      <td><p>Optional numeric, specifying sequence length,
in codons, used for filtering.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A matrix with E values for every specified subset (<code>subsets</code>,
   <code>self</code>, <code>ribosomal</code>) in columns.</p>
<p>For definition of gene expression measure (E), see
   <a href='https://bit.ly/2pGLno5'>Karlin and Mrazek (2000)</a>.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># load example DNA sequences</span>
<span class='no'>exampledir</span> <span class='kw'>&lt;-</span> <span class='fu'>system.file</span>(<span class='st'>"extdata"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"coRdon"</span>)
<span class='no'>cT</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='codonTable-class.html'>codonTable</a></span>(<span class='fu'><a href='readSet.html'>readSet</a></span>(<span class='no'>exampledir</span>))

<span class='co'># calculate E with respect to the CU</span>
<span class='co'># of ribosomal genes among the example DNA sequences</span>
<span class='no'>em</span> <span class='kw'>&lt;-</span> <span class='fu'>E</span>(<span class='no'>cT</span>, <span class='kw'>ribosomal</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
<span class='fu'>head</span>(<span class='no'>em</span>)</div><div class='output co'>#&gt;      ribosomal
#&gt; [1,] 0.9946877
#&gt; [2,] 1.0044203
#&gt; [3,] 0.9998657
#&gt; [4,] 0.9959171
#&gt; [5,] 1.0056261
#&gt; [6,] 1.0023404</div><div class='input'>
<span class='co'># calculate E distance with respect to the average CU</span>
<span class='co'># of the first 20 example DNA sequences</span>
<span class='co'># (i.e. the first half of the example DNA set)</span>
<span class='no'>em</span> <span class='kw'>&lt;-</span> <span class='fu'>E</span>(<span class='no'>cT</span>, <span class='kw'>subsets</span> <span class='kw'>=</span> <span class='fu'>list</span>(<span class='kw'>half</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='fu'>rep</span>(<span class='fl'>TRUE</span>, <span class='fl'>20</span>), <span class='fu'>rep</span>(<span class='fl'>FALSE</span>, <span class='fl'>20</span>))))

<span class='co'># alternatively, you can specify codonTable as a subset</span>
<span class='no'>halfcT</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='codonTable-class.html'>codonTable</a></span>(<span class='fu'><a href='codonTable-class.html'>codonCounts</a></span>(<span class='no'>cT</span>)[<span class='fl'>1</span>:<span class='fl'>20</span>,])
<span class='no'>em2</span> <span class='kw'>&lt;-</span> <span class='fu'>E</span>(<span class='no'>cT</span>, <span class='kw'>subsets</span> <span class='kw'>=</span> <span class='fu'>list</span>(<span class='kw'>half</span> <span class='kw'>=</span> <span class='no'>halfcT</span>))
<span class='fu'>all.equal</span>(<span class='no'>em</span>, <span class='no'>em2</span>) <span class='co'># TRUE</span></div><div class='output co'>#&gt; [1] TRUE</div><div class='input'>
<span class='co'># filtering</span>
<span class='fu'>E</span>(<span class='no'>cT</span>, <span class='kw'>ribosomal</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>filtering</span> <span class='kw'>=</span> <span class='st'>"hard"</span>, <span class='kw'>len.threshold</span> <span class='kw'>=</span> <span class='fl'>80</span>) <span class='co'># E for 9 sequences (note that,</span></div><div class='output co'>#&gt;       ribosomal
#&gt;  [1,]         1
#&gt;  [2,]         1
#&gt;  [3,]         1
#&gt;  [4,]         1
#&gt;  [5,]         1
#&gt;  [6,]         1
#&gt;  [7,]         1
#&gt;  [8,]         1
#&gt;  [9,]         1</div><div class='input'>                                          <span class='co'># accidentally, all are ribosomal)</span>
<span class='fu'>sum</span>(<span class='fu'><a href='codonTable-class.html'>getlen</a></span>(<span class='no'>cT</span>) <span class='kw'>&gt;</span> <span class='fl'>80</span>) <span class='co'># 9 sequences are longer than 80 codons</span></div><div class='output co'>#&gt; [1] 9</div><div class='input'><span class='no'>em1</span> <span class='kw'>&lt;-</span> <span class='fu'>E</span>(<span class='no'>cT</span>, <span class='kw'>ribosomal</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>filtering</span> <span class='kw'>=</span> <span class='st'>"none"</span>) <span class='co'># no filtering</span>
<span class='no'>em2</span> <span class='kw'>&lt;-</span> <span class='fu'>E</span>(<span class='no'>cT</span>, <span class='kw'>ribosomal</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>filtering</span> <span class='kw'>=</span> <span class='st'>"soft"</span>) <span class='co'># warning</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Some sequences have below-threshold length!</span></div><div class='input'><span class='fu'>all.equal</span>(<span class='no'>em1</span>, <span class='no'>em2</span>) <span class='co'># TRUE</span></div><div class='output co'>#&gt; [1] TRUE</div><div class='input'>
<span class='co'># options for genetic code</span>
<span class='no'>em</span> <span class='kw'>&lt;-</span> <span class='fu'>E</span>(<span class='no'>cT</span>, <span class='kw'>ribosomal</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
        <span class='kw'>stop.rm</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>) <span class='co'># don't use stop codons in calculation</span>
<span class='no'>em</span> <span class='kw'>&lt;-</span> <span class='fu'>E</span>(<span class='no'>cT</span>, <span class='kw'>ribosomal</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
        <span class='kw'>alt.init</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>) <span class='co'># don't use alternative start codons</span>
<span class='no'>em</span> <span class='kw'>&lt;-</span> <span class='fu'>E</span>(<span class='no'>cT</span>, <span class='kw'>ribosomal</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
        <span class='kw'>id_or_name2</span> <span class='kw'>=</span> <span class='st'>"2"</span>) <span class='co'># use different genetic code, for help</span>
                           <span class='co'># see `?Biostrings::GENETIC_CODE`</span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://bioinfo.hr/members/'>Anamaria Elek</a>, <a href='http://bioinfo.hr/maja-kuzman/'>Maja Kuzman</a>, <a href='http://bioinfo.hr/kristian-vlahovicek/'>Kristian Vlahovicek</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

